core/clipboard/block_paster.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {BlockSvg} from '../block_svg.js';
+import * as registry from './registry.js';
+import {ICopyData} from '../interfaces/i_copyable.js';
+import {IPaster} from '../interfaces/i_paster.js';
+import {State, append} from '../serialization/blocks.js';
+import {Coordinate} from '../utils/coordinate.js';
+import {WorkspaceSvg} from '../workspace_svg.js';
+import * as eventUtils from '../events/utils.js';
+import {config} from '../config.js';
+
+export class BlockPaster implements IPaster<BlockCopyData, BlockSvg> {
+  static TYPE = 'block';
+
+  paste(
+    copyData: BlockCopyData,
+    workspace: WorkspaceSvg,
+    coordinate?: Coordinate,
+  ): BlockSvg | null {
+    if (!workspace.isCapacityAvailable(copyData.typeCounts!)) return null;
+
+    if (coordinate) {
+      copyData.blockState['x'] = coordinate.x;
+      copyData.blockState['y'] = coordinate.y;
+    }
+
+    eventUtils.disable();
+    let block;
+    try {
+      block = append(copyData.blockState, workspace) as BlockSvg;
+      moveBlockToNotConflict(block);
+    } finally {
+      eventUtils.enable();
+    }
+
+    if (!block) return block;
+
+    if (eventUtils.isEnabled() && !block.isShadow()) {
+      eventUtils.fire(new (eventUtils.get(eventUtils.BLOCK_CREATE))(block));
+    }
+    block.select();
+    return block;
+  }
+}
+
+/**
+ * Moves the given block to a location where it does not: (1) overlap exactly
+ * with any other blocks, or (2) look like it is connected to any other blocks.
+ *
+ * Exported for testing.
+ *
+ * @param block The block to move to an unambiguous location.
+ * @internal
+ */
+export function moveBlockToNotConflict(block: BlockSvg) {
```

Comment:
`bumpNeighbors` always moves whichever block has the inferior connection :/ We want to make sure we're only moving the block that's being pasted.

Wrt code reuse: we're using `connection.closest` instead of `connection.neighbours`. They both take in a snap radius and do similar things.